File: java/AvoidHadoukenIfs.java
// Avoid Hadouken IFs
// Avoid nested IFs (Hadouken IFs)
// Solution: Early Returns, Switch-Cases

public class AvoidHadoukenIfs {
    static class Order {
        private boolean isValid = true;
        private boolean isInStock = true;
        private boolean paymentSuccess = true;

        public boolean valid() {
            return isValid;
        }

        public boolean inStock() {
            return isInStock;
        }

        public boolean paymentSuccessful() {
            return paymentSuccess;
        }
    }

    // Before
    public static void processOrderBefore(Order order) {
        if (order.valid()) {
            if (order.inStock()) {
                if (order.paymentSuccessful()) {
                    System.out.println("Order processed successfully!");
                } else {
                    System.out.println("Payment failed.");
                }
            } else {
                System.out.println("Item is out of stock.");
            }
        } else {
            System.out.println("Order is invalid.");
        }
    }

    // After
    public static void processOrder(Order order) {
        if (!order.valid()) {
            System.out.println("Order is invalid.");
            return;
        }
        
        if (!order.inStock()) {
            System.out.println("Item is out of stock.");
            return;
        }
        
        if (!order.paymentSuccessful()) {
            System.out.println("Payment failed.");
            return;
        }

        System.out.println("Order processed successfully!");
    }

    public static void main(String[] args) {
        Order order = new Order();
        
        processOrderBefore(order);
        System.out.println();
        processOrder(order);
    }
}

File: rust/avoid_hadouken_ifs.rs
// Avoid Hadouken IFs
// Avoid nested IFs (Hadouken IFs)
// Solution: Early Returns, Switch-Cases

struct Order {
    is_valid: bool,
    is_in_stock: bool,
    payment_success: bool,
}

impl Order {
    fn new() -> Self {
        Order {
            is_valid: true,
            is_in_stock: true,
            payment_success: true,
        }
    }

    fn valid(&self) -> bool {
        self.is_valid
    }

    fn in_stock(&self) -> bool {
        self.is_in_stock
    }

    fn payment_successful(&self) -> bool {
        self.payment_success
    }
}

// Before
fn process_order_before(order: &Order) {
    if order.valid() {
        if order.in_stock() {
            if order.payment_successful() {
                println!("Order processed successfully!");
            } else {
                println!("Payment failed.");
            }
        } else {
            println!("Item is out of stock.");
        }
    } else {
        println!("Order is invalid.");
    }
}

// After
fn process_order(order: &Order) {
    if !order.valid() {
        println!("Order is invalid.");
        return;
    }
    
    if !order.in_stock() {
        println!("Item is out of stock.");
        return;
    }
    
    if !order.payment_successful() {
        println!("Payment failed.");
        return;
    }

    println!("Order processed successfully!");
}

fn main() {
    let order = Order::new();
    
    process_order_before(&order);
    println!();
    process_order(&order);
}
